### PR TITLE
[TACHYON-964] Start Tachyon using YARN

### PR DESCRIPTION
**Pull request by apc999**
JIRA: https://tachyon.atlassian.net/browse/TACHYON-964

This PR aims to enable starting Tachyon by YARN.

### CODE REVIEW

**Comment by calvinjia** [comment]
Perhaps mesos and yarn should be in separate sub modules? Perhaps a follow up after this is reviewed/merged.

**Comment by apc999** [comment]
@calvinjia , yes, there are definitely some refactoring work can and should be done between mesos and yarn parts. And let's aim that in followup PRs.

File: integration/yarn/src/main/java/tachyon/yarn/Client.java [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+    mAppContext = app.getApplicationSubmissionContext();
+    setupApplicationSubmissionContext();
+
+    //
+    // Submit the application to the applications manager.
+    //
+    // Ignore the response as either a valid response object is returned on success
+    // or an exception thrown to denote some form of a failure
+    mAppId = mAppContext.getApplicationId();
+    System.out.println("Submitting application of id " + mAppId + " to ASM");
```
Comment (uronce-cc) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: What is "ASM"?

File: integration/yarn/src/main/java/tachyon/yarn/Client.java [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+    mYarnClient.submitApplication(mAppContext);
+  }
+
+  // Checks if the cluster has enough resource to launch application master
+  private void checkClusterResource(GetNewApplicationResponse appResponse) {
+    int maxMem = appResponse.getMaximumResourceCapability().getMemory();
+    int maxVCores = appResponse.getMaximumResourceCapability().getVirtualCores();
+
+    Preconditions.checkArgument(mAmMemory <= maxMem,
+        "AM memory specified above max threshold of cluster, specified=" + mAmMemory + ", max="
```
Comment (uronce-cc) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: Will users understand "AM" as "app master", is that a concept in yarn?

File: integration/yarn/src/main/java/tachyon/yarn/Client.java [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+            + maxMem);
+    Preconditions.checkArgument(mAmVCores <= maxVCores,
+        "AM virtual cores specified above max threshold of cluster, specified=" + mAmVCores
```
Comment (uronce-cc) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: ditto

File: integration/bin/tachyon-worker-yarn.sh [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+SCRIPT_DIR="$(cd "$(dirname "$0")"; pwd)"
+source "${SCRIPT_DIR}/common.sh"
+TACHYON_WORKER_JAVA_OPTS="${TACHYON_WORKER_JAVA_OPTS:-${TACHYON_JAVA_OPTS}}"
+
+echo Mount ramdisk
```
Comment (jsimsa) [inline integration/bin/tachyon-worker-yarn.sh]: It makes sense to do the same in tachyon-worker-mesos.sh (and remove the explicit mounting from vagrant).

File: integration/pom.xml [hunk 5]
```diff
@@ -1,57 +1,24 @@
...
   <parent>
-    <artifactId>tachyon-parent</artifactId>
     <groupId>org.tachyonproject</groupId>
+    <artifactId>tachyon-parent</artifactId>
     <version>0.8.0-SNAPSHOT</version>
   </parent>
-  <name>Tachyon Integration</name>
   <artifactId>tachyon-integration</artifactId>
+  <packaging>pom</packaging>
+  <name>Tachyon Integration with Other Systems</name>
```
Comment (jsimsa) [inline integration/pom.xml]: nit: I would just say "Tachyon Integration"

File: integration/pom.xml [hunk 6]
```diff
@@ -1,57 +1,24 @@
...
-  <build>
-    <resources>
-      <resource>
-        <directory>src/main/resources</directory>
-        <filtering>true</filtering>
-      </resource>
-    </resources>
-  </build>
-</project>
+</project>
```
Comment (jsimsa) [inline integration/pom.xml]: missing new line

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+  @Override
+  public void onError(Throwable t) {}
+
+  @Override
+  public float getProgress() {
+    return 0;
+  }
+
+  public void start() throws IOException, YarnException {
+    // create a client to talk to NM
```
Comment (jsimsa) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: expand NM to node manager

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+    mNMClient = NMClient.createNMClient();
+    mNMClient.init(mYarnConf);
+    mNMClient.start();
+
+    // Create a client to talk to the RM
```
Comment (jsimsa) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: expand RM to resource manager

File: integration/yarn/src/main/java/tachyon/yarn/Client.java [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+  }
+
+  /**
+   * Parses command line options.
+   *
+   * @param args Parsed command line options
+   * @return Whether the parseArgs was successful to run the client
+   * @throws ParseException
+   */
+  private boolean parseArgs(String[] args) throws ParseException {
```
Comment (jsimsa) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: This would be so much easier in Go :-)

File: integration/yarn/src/main/java/tachyon/yarn/Client.java [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+   * Submits an application to the ResourceManager to run ApplicationMaster.
+   *
+   * The stable Yarn API provides a convenience method (YarnClient#createApplication) for creating
+   * applications and setting up the application submission context. This was not available in the
+   * alpha API.
+   */
+  private void submitApplication() throws YarnException, IOException {
+    // TODO: setup credential
+
+    //
```
Comment (jsimsa) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: Do we need these empty commented lines?

File: integration/yarn/src/main/java/tachyon/yarn/Client.java (hunk 1 above)
Comment (jsimsa) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: Expand to whathever ASM is an abbreviation for.

File: integration/yarn/src/main/java/tachyon/yarn/Client.java (hunk 2 above)
Comment (jsimsa) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: ditto

File: integration/yarn/src/main/java/tachyon/yarn/Client.java (hunk 3 above)
Comment (jsimsa) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: ditto

File: integration/yarn/src/main/java/tachyon/yarn/Client.java [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+
+    // Set the priority for the application master
+    mAppContext.setPriority(Priority.newInstance(mAmPriority));
+  }
+
+  /**
+   * Monitors the submitted application for completion.
+   *
+   * @return true if application completed successfully
+   * @throws YarnException
```
Comment (jsimsa) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: add exception case description

File: integration/yarn/src/main/java/tachyon/yarn/Client.java [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+   * @throws IOException
```
Comment (jsimsa) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: ditto

File: integration/yarn/src/main/java/tachyon/yarn/Client.java [hunk 13]
```diff
@@ -0,0 +1,388 @@
...
+    }
+    System.err.println("Application failed to complete");
+    System.exit(2);
+  }
+
+  /**
+   * Main run function for the client
+   *
+   * @return true if application completed successfully
+   * @throws IOException
```
Comment (jsimsa) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: add exception case description

File: integration/yarn/src/main/java/tachyon/yarn/Client.java [hunk 14]
```diff
@@ -0,0 +1,388 @@
...
+   * @throws YarnException
```
Comment (jsimsa) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: ditto

File: integration/yarn/src/main/java/tachyon/yarn/Client.java [hunk 15]
```diff
@@ -0,0 +1,388 @@
...
+  private void printUsage() {
+    new HelpFormatter().printHelp("Client", mOptions);
+  }
+
+  /**
+   * Parses command line options.
+   *
+   * @param args Parsed command line options
+   * @return Whether the parseArgs was successful to run the client
+   * @throws ParseException
```
Comment (jsimsa) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: ditto

File: integration/yarn/src/main/java/tachyon/yarn/CommandBuilder.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+package tachyon.yarn;
+
+import java.util.Vector;
+
+import com.google.common.base.Preconditions;
+
+/**
+ * A shell command.
+ *
+ * TODO: to share this class into util.
```
Comment (jsimsa) [inline integration/yarn/src/main/java/tachyon/yarn/CommandBuilder.java]: TODO(bin)

File: integration/yarn/src/main/java/tachyon/yarn/CommandBuilder.java [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+package tachyon.yarn;
+
+import java.util.Vector;
+
+import com.google.common.base.Preconditions;
+
+/**
+ * A shell command.
```
Comment (jsimsa) [inline integration/yarn/src/main/java/tachyon/yarn/CommandBuilder.java]: This is not very descriptive.

File: integration/yarn/src/main/java/tachyon/yarn/Client.java [hunk 18]
```diff
@@ -0,0 +1,388 @@
...
+
+  /**
+   * Submits an application to the ResourceManager to run ApplicationMaster.
+   *
+   * The stable Yarn API provides a convenience method (YarnClient#createApplication) for creating
+   * applications and setting up the application submission context. This was not available in the
+   * alpha API.
+   */
+  private void submitApplication() throws YarnException, IOException {
+    // TODO: setup credential
```
Comment (jsimsa) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: TODO(bin)

**Comment by jsimsa** [comment]
Left you some minor comments. I don't think your changes to the Mesos sources location require any changes to the vagrant deployment, but it would be great if you could test that it still works.

Also, do you have a write up the describes how to test / deploy the integration (similar to the one I have on Mesos)?

File: integration/mesos/pom.xml [hunk 19]
```diff
@@ -0,0 +1,57 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (yupeng9) [inline integration/mesos/pom.xml]: Perhaps we should add this mesos-related file in a separate PR?

Comment (jsimsa) [inline integration/mesos/pom.xml]: This use to be integration/pom.xml prior to this PR. Git just does not understand that this is a rename operation.

File: integration/bin/tachyon-master-yarn.sh [hunk 20]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+SCRIPT_DIR="$(cd "$(dirname "$0")"; pwd)"
+source "${SCRIPT_DIR}/common.sh"
+TACHYON_MASTER_JAVA_OPTS="${TACHYON_MASTER_JAVA_OPTS:-${TACHYON_JAVA_OPTS}}"
+
+mkdir -p "${TACHYON_LOGS_DIR}"
+
+"${JAVA}" -cp "${CLASSPATH}" \
+  ${TACHYON_MASTER_JAVA_OPTS} \
```
Comment (calvinjia) [inline integration/bin/tachyon-master-yarn.sh]: Should this go last? Otherwise the user would not be able to modify the options below.

File: integration/bin/tachyon-worker-yarn.sh [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+${TACHYON_HOME}/bin/tachyon-mount.sh SudoMount
```
Comment (calvinjia) [inline integration/bin/tachyon-worker-yarn.sh]: Is sudomount required or would mount be sufficient?

File: integration/bin/tachyon-worker-yarn.sh [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+
+mkdir -p "${TACHYON_LOGS_DIR}"
+
+echo Launch Tachyon worker
+"${JAVA}" -cp "${CLASSPATH}" \
+  ${TACHYON_WORKER_JAVA_OPTS}  \
```
Comment (calvinjia) [inline integration/bin/tachyon-worker-yarn.sh]: Same comment about ordering.

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java [hunk 23]
```diff
@@ -0,0 +1,280 @@
...
+ * ApplicationMaster on an allocated container. The ApplicationMaster communicates with YARN
+ * cluster, and handles application execution. It performs operations in an asynchronous fashion.
+ */
+public final class ApplicationMaster implements AMRMClientAsync.CallbackHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  // Parameters sent from Client
+  private final int mMasterCpu;
+  private final int mWorkerCpu;
+  private final int mMasterMem;
```
Comment (calvinjia) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: Maybe specify the units (ie `mb`)?

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java [hunk 24]
```diff
@@ -0,0 +1,280 @@
...
+    List<String> commands = Lists.newArrayList(command);
+
+    for (Container container : containers) {
+      try {
+        // Launch container by create ContainerLaunchContext
+        ContainerLaunchContext ctx = Records.newRecord(ContainerLaunchContext.class);
+        ctx.setCommands(commands);
+        LOG.info("Launching container {} for Tachyon master on {} ", container.getId(),
+            container.getNodeHttpAddress());
+        LOG.info("--------- with master command: " + commands);
```
Comment (calvinjia) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: Could these two be combined into one log message?

File: pom.xml [hunk 25]
```diff
@@ -537,5 +536,22 @@
...
+      <id>mesos</id>
+      <modules>
+        <module>integration/mesos</module>
+      </modules>
+    </profile>
+
+    <profile>
+      <id>yarn</id>
+      <modules>
+        <module>integration/yarn</module>
```
Comment (calvinjia) [inline pom.xml]: Maybe add a comment for the hadoop dependency?

File: integration/yarn/src/main/java/tachyon/yarn/Client.java (hunk 1 above)
Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: Done.

File: integration/yarn/src/main/java/tachyon/yarn/Client.java (hunk 2 above)
Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: Done.

File: integration/bin/tachyon-worker-yarn.sh (hunk 4 above)
Comment (apc999) [inline integration/bin/tachyon-worker-yarn.sh]: Will do this in a separate PR to cleanup mesos related

File: integration/pom.xml (hunk 5 above)
Comment (apc999) [inline integration/pom.xml]: done.

File: integration/pom.xml (hunk 6 above)
Comment (apc999) [inline integration/pom.xml]: done.

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java (hunk 8 above)
Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: done

File: integration/yarn/src/main/java/tachyon/yarn/Client.java (hunk 10 above)
Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: done.

File: integration/yarn/src/main/java/tachyon/yarn/Client.java (hunk 3 above)
Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: done

File: integration/yarn/src/main/java/tachyon/yarn/Client.java (hunk 11 above)
Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: added minimal explanation as that is what I could get from YARN api documentation.

File: integration/yarn/src/main/java/tachyon/yarn/Client.java (hunk 12 above)
Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: added minimal explanation as that is what I could get from YARN api documentation.

File: integration/yarn/src/main/java/tachyon/yarn/CommandBuilder.java (hunk 16 above)
Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/CommandBuilder.java]: done.

File: integration/yarn/src/main/java/tachyon/yarn/CommandBuilder.java (hunk 17 above)
Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/CommandBuilder.java]: updated javadoc.

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java (hunk 7 above)
Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: done

File: integration/yarn/src/main/java/tachyon/yarn/Client.java (hunk 13 above)
Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: done---though it is very vague because that 's all I could get from YARN documentation.

File: integration/yarn/src/main/java/tachyon/yarn/Client.java (hunk 14 above)
Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: done---though it is very vague because that 's all I could get from YARN documentation.

File: integration/yarn/src/main/java/tachyon/yarn/Client.java (hunk 15 above)
Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: Fixed.

File: integration/yarn/src/main/java/tachyon/yarn/Client.java (hunk 18 above)
Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: done.

File: integration/bin/tachyon-worker-yarn.sh (hunk 21 above)
Comment (apc999) [inline integration/bin/tachyon-worker-yarn.sh]: It is required, unfortunately.

File: integration/bin/tachyon-master-yarn.sh (hunk 20 above)
Comment (apc999) [inline integration/bin/tachyon-master-yarn.sh]: Good catch. Fixed.

File: integration/bin/tachyon-worker-yarn.sh (hunk 22 above)
Comment (apc999) [inline integration/bin/tachyon-worker-yarn.sh]: Fixed.

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java (hunk 23 above)
Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: done.

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java (hunk 24 above)
Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: done

File: pom.xml (hunk 25 above)
Comment (apc999) [inline pom.xml]: good call. Added.

**Comment by apc999** [comment]
@jsimsa I will add the description in a markdown and cleanup the mesos shell in a followup PR.

File: integration/yarn/src/main/java/tachyon/yarn/Client.java [hunk 26]
```diff
@@ -0,0 +1,378 @@
...
+
+  /**
+   * Submits an application to the ResourceManager to run ApplicationMaster.
+   *
+   * The stable Yarn API provides a convenience method (YarnClient#createApplication) for creating
+   * applications and setting up the application submission context. This was not available in the
+   * alpha API.
+   */
+  private void submitApplication() throws YarnException, IOException {
+    // TODO(binfan): setup credential
```
Comment (calvinjia) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: Will it be in this PR?

File: integration/yarn/src/main/java/tachyon/yarn/Client.java [hunk 27]
```diff
@@ -0,0 +1,378 @@
...
+    while (true) {
+      // Check app status every 10 second.
+      CommonUtils.sleepMs(10 * Constants.SECOND_MS);
+      Date date = new Date();
+      String timeDateStr = dateFormat.format(date);
+
+      // Get application report for the appId we are interested in
+      ApplicationReport report = mYarnClient.getApplicationReport(mAppId);
+
+      System.out.println(timeDateStr + " Got application report from ASM for appId="
```
Comment (calvinjia) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: Would it be easier to use a logger?

File: integration/yarn/src/main/java/tachyon/yarn/Client.java (hunk 26 above)
Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: no. it will be in a much broader context I believe

File: integration/yarn/src/main/java/tachyon/yarn/Client.java (hunk 27 above)
Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/Client.java]: I suspect people will run this using commandline, so using `System.out.println` to get message to console through this file.

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java [hunk 28]
```diff
@@ -0,0 +1,281 @@
...
+    for (Container container : containers) {
+      try {
+        // Launch container by create ContainerLaunchContext
+        ContainerLaunchContext ctx = Records.newRecord(ContainerLaunchContext.class);
+        ctx.setCommands(commands);
+        LOG.info("Launching container {} for Tachyon master on {} with master command: {}",
+            container.getId(), container.getNodeHttpAddress(), commands);
+        mNMClient.startContainer(container, ctx);
+        String containerUri = container.getNodeHttpAddress(); // in the form of 1.2.3.4:8042
+        mMasterContainerNetAddress = containerUri.split(":")[0];
```
Comment (calvinjia) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: Since you can get the master address here, does it matter which node it gets launched on? (do we need `mMasterAddress` passed in still)?

Comment (calvinjia) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: Oh I see, its mostly for ease of use from the user side.

**Comment by calvinjia** [comment]
LGTM

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java [hunk 29]
```diff
@@ -0,0 +1,281 @@
...
+    }
+
+    // Resource requirements for master containers
+    Resource workerResource = Records.newRecord(Resource.class);
+    workerResource.setMemory(mWorkerMemInMB);
+    workerResource.setVirtualCores(mWorkerCpu);
+
+    // Make container requests for workers to ResourceManager
+    for (int i = 0; i < mNumWorkers; i ++) {
+      ContainerRequest containerAsk = new ContainerRequest(workerResource, null /* any hosts */,
```
Comment (GraceH) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: One quick question about this line. since location hint is ANY.  how can we ensure there won't be two workers on the same physical node?

Comment (apc999) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: @GraceH , Thanks for the finding! Yes as you mentioned, currently two workers will not work if they are assigned to the same container. I will submit a PR to fix this as in Mesos imeplementation where we try to make sure two workers do not assigned to the same container!
